File: apps/web/src/lib/api/ai-tools.ts
import type { AiRiskTier, AiTool } from "@/lib/mock-data/ai-os";

export type ApiAiTool = {
  id: string;
  company_id: string;
  name: string;
  agent: string;
  category: string;
  risk: string;
  description: string;
  created_at: string;
  updated_at: string;
};

export type ApiAiToolListResponse = {
  data: ApiAiTool[];
  meta: { count: number };
};

export type ApiAiToolResponse = {
  data: ApiAiTool;
};

export type AiToolListParams = {
  category?: string;
  agent?: string;
  risk?: AiRiskTier;
};

export function apiAiToolToTool(row: ApiAiTool): AiTool {
  return {
    id: row.id,
    name: row.name,
    agent: row.agent,
    category: row.category,
    risk: row.risk as AiRiskTier,
    description: row.description,
  };
}

export function buildAiToolQuery(params?: AiToolListParams): string {
  if (!params) return "";
  const q = new URLSearchParams();
  if (params.category) q.set("category", params.category);
  if (params.agent) q.set("agent", params.agent);
  if (params.risk) q.set("risk", params.risk);
  const s = q.toString();
  return s ? `?${s}` : "";
}

File: apps/api/app/routers/ai_tools.py
from __future__ import annotations

from typing import Optional

from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy.orm import Session

from app.database import get_db
from app.models.ai_tool import AiTool
from app.schemas.ai_tool import (
    AiToolListMeta,
    AiToolListResponse,
    AiToolRead,
    AiToolResponse,
)

router = APIRouter(prefix="/tools", tags=["ai-tools"])


def _to_read(row: AiTool) -> AiToolRead:
    return AiToolRead(
        id=row.id,
        company_id=row.company_id,
        name=row.name,
        agent=row.agent,
        category=row.category,
        risk=row.risk,
        description=row.description,
        created_at=row.created_at,
        updated_at=row.updated_at,
    )


@router.get("", response_model=AiToolListResponse)
def list_tools(
    db: Session = Depends(get_db),
    category: Optional[str] = Query(default=None),
    agent: Optional[str] = Query(default=None),
    risk: Optional[str] = Query(default=None),
) -> AiToolListResponse:
    query = db.query(AiTool).order_by(AiTool.category.asc(), AiTool.name.asc())

    if category:
        query = query.filter(AiTool.category == category)
    if agent:
        query = query.filter(AiTool.agent == agent)
    if risk:
        query = query.filter(AiTool.risk == risk)

    rows = query.all()
    data = [_to_read(row) for row in rows]
    return AiToolListResponse(
        data=data,
        meta=AiToolListMeta(count=len(data)),
    )


@router.get("/{tool_id}", response_model=AiToolResponse)
def get_tool(tool_id: str, db: Session = Depends(get_db)) -> AiToolResponse:
    row = db.get(AiTool, tool_id)
    if not row:
        raise HTTPException(status_code=404, detail="Tool not found")
    return AiToolResponse(data=_to_read(row))
